File: apps/web/src/types/auditLog.ts
/**
 * Type definitions for Audit Log feature
 * Story 2.13 Task 14: Implement Audit Log
 */

/**
 * Audit log entry DTO from backend
 * Represents a single audit trail record for admin CRUD operations
 */
export interface AuditLogDto {
  /**
   * Unique identifier for the audit log entry
   */
  id: string

  /**
   * Timestamp when the action occurred (UTC, ISO 8601 format)
   */
  timestamp: string

  /**
   * User identifier who performed the action (email or username)
   */
  userId: string

  /**
   * Action performed: "Create", "Update", "Delete", "Import"
   */
  action: 'Create' | 'Update' | 'Delete' | 'Import'

  /**
   * Entity type affected: "Model", "Benchmark", "BenchmarkScore"
   */
  entityType: 'Model' | 'Benchmark' | 'BenchmarkScore'

  /**
   * Unique identifier of the affected entity
   */
  entityId: string

  /**
   * JSON string representing the entity BEFORE the operation
   * NULL for Create actions (no previous state)
   */
  oldValues: string | null

  /**
   * JSON string representing the entity AFTER the operation
   * NULL for Delete actions (entity no longer exists)
   */
  newValues: string | null
}

/**
 * Audit log filter parameters for API queries
 */
export interface AuditLogFilters {
  /**
   * Filter by user identifier (email or username)
   */
  userId?: string

  /**
   * Filter by entity type
   */
  entityType?: string

  /**
   * Filter by action
   */
  action?: string

  /**
   * Filter for audit logs after this date (inclusive, ISO 8601 format)
   */
  startDate?: string

  /**
   * Filter for audit logs before this date (inclusive, ISO 8601 format)
   */
  endDate?: string

  /**
   * Page number (1-indexed)
   */
  page?: number

  /**
   * Page size (default: 20, max: 100)
   */
  pageSize?: number
}

/**
 * Pagination metadata from backend
 */
export interface PaginationMetadata {
  currentPage: number
  pageSize: number
  totalItems: number
  totalPages: number
  hasNextPage: boolean
  hasPreviousPage: boolean
}

/**
 * Paginated audit log response
 */
export interface PagedAuditLogResult {
  items: AuditLogDto[]
  pagination: PaginationMetadata
}

/**
 * API response wrapper for audit logs
 */
export interface AuditLogResponse {
  data: PagedAuditLogResult
  meta: {
    timestamp: string
  }
}

File: services/backend/LlmTokenPrice.Domain/Entities/AuditLog.cs
namespace LlmTokenPrice.Domain.Entities;

/// <summary>
/// Represents an audit log entry capturing all admin CRUD operations for compliance and traceability.
/// This is a pure domain entity with no infrastructure dependencies (EF attributes, data annotations, etc.).
/// </summary>
/// <remarks>
/// Entity configuration via Fluent API in Infrastructure layer (AuditLogConfiguration.cs).
/// Table name: audit_logs (snake_case per PostgreSQL conventions).
/// Supports immutable append-only logging pattern - records are NEVER updated or deleted.
/// </remarks>
public class AuditLog
{
    /// <summary>
    /// Unique identifier for the audit log entry (UUID/GUID for distributed system compatibility).
    /// </summary>
    public Guid Id { get; set; }

    /// <summary>
    /// Timestamp when the action occurred (UTC).
    /// Indexed for date range filtering and chronological queries.
    /// </summary>
    public DateTime Timestamp { get; set; }

    /// <summary>
    /// User identifier who performed the action (email or username from JWT claims).
    /// Indexed for filtering by user.
    /// </summary>
    public string UserId { get; set; } = null!;

    /// <summary>
    /// Action performed: "Create", "Update", "Delete", "Import".
    /// Constrained to enum values in infrastructure layer.
    /// Indexed for action type filtering.
    /// </summary>
    public string Action { get; set; } = null!;

    /// <summary>
    /// Entity type affected: "Model", "Benchmark", "BenchmarkScore".
    /// Indexed for entity type filtering.
    /// </summary>
    public string EntityType { get; set; } = null!;

    /// <summary>
    /// Unique identifier of the affected entity (e.g., Model.Id, Benchmark.Id).
    /// Indexed for entity-level audit trail queries.
    /// </summary>
    public Guid EntityId { get; set; }

    /// <summary>
    /// JSON-serialized representation of the entity BEFORE the operation.
    /// NULL for Create actions (no previous state).
    /// Stored as JSONB in PostgreSQL for queryability and compression.
    /// </summary>
    public string? OldValues { get; set; }

    /// <summary>
    /// JSON-serialized representation of the entity AFTER the operation.
    /// NULL for Delete actions (entity no longer exists).
    /// Stored as JSONB in PostgreSQL for queryability and compression.
    /// </summary>
    public string? NewValues { get; set; }
}
